addons/membership/models/membership.py

```diff
@@ -20,50 +15,43 @@
-            if invoice.invoice_line_ids:
-                list_membership_line += member_line_obj.search(cr, uid, [('account_invoice_line', 'in', [ l.id for l in invoice.invoice_line_ids])], context=context)
-        return list_membership_line
-
-    def _check_membership_date(self, cr, uid, ids, context=None):
-        """Check if membership product is not in the past """
-
-        cr.execute('''
-         SELECT MIN(ml.date_to - ai.date_invoice)
-             FROM membership_membership_line ml
-             JOIN account_invoice_line ail ON (
-                ml.account_invoice_line = ail.id
-                )
-            JOIN account_invoice ai ON (
-            ai.id = ail.invoice_id)
-            WHERE ml.id IN %s''', (tuple(ids),))
-        res = cr.fetchall()
-        for r in res:
-            if r[0] and r[0] < 0:
-                return False
-        return True
+    partner = fields.Many2one('res.partner', string='Partner', ondelete='cascade', index=True)
+    membership_id = fields.Many2one('product.product', string="Membership", required=True)
+    date_from = fields.Date(string='From', readonly=True)
+    date_to = fields.Date(string='To', readonly=True)
+    date_cancel = fields.Date(string='Cancel date')
+    date = fields.Date(string='Join Date',
+        help="Date on which member has joined the membership")
+    member_price = fields.Float(string='Membership Fee',
+        digits_compute= dp.get_precision('Product Price'), required=True,
+        help='Amount for the membership')
+    account_invoice_line = fields.Many2one('account.invoice.line', string='Account Invoice line', readonly=True)
+    account_invoice_id = fields.Many2one('account.invoice', related='account_invoice_line.invoice_id', string='Invoice', readonly=True)
+    company_id = fields.Many2one('res.company', related='account_invoice_line.invoice_id.company_id', string="Company", readonly=True, store=True)
+    state = fields.Selection(STATE, compute='_compute_state', string='Membership Status', store=True,
+        help="It indicates the membership status.\n"
+             "-Non Member: A member who has not applied for any membership.\n"
+             "-Cancelled Member: A member who has cancelled his membership.\n"
+             "-Old Member: A member whose membership date has expired.\n"
+             "-Waiting Member: A member who has applied for the membership and whose invoice is going to be created.\n"
+             "-Invoiced Member: A member whose invoice has been created.\n"
+             "-Paid Member: A member who has paid the membership amount.")
 
-    def _state(self, cr, uid, ids, name, args, context=None):
+    @api.depends('account_invoice_line', 'account_invoice_line.invoice_id',
+                 'account_invoice_line.invoice_id.state',
+                 'account_invoice_line.invoice_id.payment_ids',
+                 'account_invoice_line.invoice_id.payment_ids.invoice_ids')
```

Comment:
I think that it's false because: 
- You don't use the second trigger method and model
- For the first trigger condition, you use too many fields. 

I think that this is sufficient. 
``` python
    @api.depends('account_invoice_line.invoice_id.state','partner.membership_state')
```